File: fullflu/binary_search_tree.cpp
#include <vector>
#include <stack>
#include <stdio.h>
#include <iostream>
#include <string>
#include <algorithm>

#include <typeinfo>

using namespace std;

struct Node
{
	int key;
	Node *parent, *left, *right;
};
Node *root,*NIL;

// print in preorder
void preorder(Node *T_1){
	if(T_1 == NIL){
		return;
	}
	cout << " " << T_1->key;
	preorder(T_1->left);
	preorder(T_1->right);
}

// print in inorder
void inorder(Node *T_1){
	if(T_1 == NIL){
		return;
	}
	inorder(T_1->left);
	cout << " " << T_1->key;
	inorder(T_1->right);
}

// insert node
void insert(int k){
	Node *y = NIL;
	Node *x = root;
	Node *z;

	z = (Node *)malloc(sizeof(Node));
	z->key = k;
	z->left = NIL;
	z->right = NIL;
	//cout << endl;

	while( x!= NIL){
		//cout << "aa " << endl;
		y = x;
		if(z -> key < x->key){
			x = x->left;
		}
		else{
			x = x->right;
		}
	}
	z->parent = y;

	if(y == NIL){
		root = z;
	}
	else{
		if(z->key < y->key){
			y->left = z;
		}
		else{
			y->right = z;
		}
	}

}

// honest recursive search leads to TimeLimitExceeding
bool find_flg = false;
void find_recursive(Node *x,int k){
	if(find_flg == true || x == NIL) return;
	if(x->key == k){
		find_flg = true;
		return;
	}
	find_recursive(x->left,k);
	find_recursive(x->right,k);
}

// finding node
Node *find_tree(Node *x, int k){
	while(x != NIL && k != x->key){
		if(k < x->key) x = x->left;
		else x = x->right;
	}
	return x;
}

// find a leftmost child
Node *treeMinimum(Node *target){
	while(target->left != NIL){
		target = target->left;
	}
	return target;
}

// find a next successor
Node *get_treeSuccessor(Node *target){
	// check if the target has a right child (this must be true in the task of delete_tree)
	if(target->right != NIL){
		return treeMinimum(target->right);
	}
	cout << "this must not be shown" << endl;
	// ignore at the task of delete_tree
	Node *parent = target->parent;
	while(parent != NIL && target == parent->right){
		target = parent;
		parent = parent->parent;
	}
	return parent;
}

// delete node
void delete_tree(int k){
	Node *target = find_tree(root,k);
	Node *removing;
	Node *child;
	// check if tree includes k
	if(target == NIL){
		cout << "cannot delete:" << k << endl;
		return;
	}
	// decide which node will be removed explicitly
	// if the target has at most 1 child, remove the target
	// else, remove Successor of the target
	else if(target->left == NIL || target->right == NIL)
		removing = target;
	else{
		removing = get_treeSuccessor(target);
		//cout << removing->key << endl;
	}


	// define a transferred child of the removing node
	// if removing node has a left child (Not NIL), it is the child
	// else, a right child (including NIL) is the child
	if(removing->left != NIL)
		child = removing->left;
	else
		child = removing->right;

	// set the parent of the transferred child
	if(child != NIL)
		child->parent = removing->parent;

	// set the root of the tree or the child of the parent of removing node
	if(removing->parent == NIL)
		root = child;
	else if(removing == removing->parent->left)
		removing->parent->left = child;
	else
		removing->parent->right = child;


	if(target != removing)
		target->key = removing->key;

	// delete removing node
	free(removing);
}

// implementation of binary search tree
void binary_search_tree(){
	int n,k;
	Node *ans;
	string com;
	cin >> n;

	for(int i = 0; i < n; i++){
		cin >> com;
		if(com == "insert"){
			cin >> k;
			//cout << typeid(k).name() << endl;
			insert(k);
		}
		else if(com == "print"){
			inorder(root);
			cout << endl;
			preorder(root);
			cout << endl;
		}
		else if(com == "find"){
			cin >> k;
			ans = find_tree(root,k);
			if(ans != NIL){
				cout << "yes";
				//find_flg = false;
				cout << endl;
			}
			else{
				cout << "no";
				cout << endl;
			}
		}
		else if(com == "delete"){
			cin >> k;
			delete_tree(k);
		}

	}
}

int main(){
	binary_search_tree();
	return 0;
}

File: fullflu/binary_search_tree.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import sys

# BinaryTree: init, insert, find, and delete.
# inefficient
# Time limit exceeding at the last test (of BinaryTree 2 and 3).

class BinaryTree(object):
    def __init__(self):
        self.tree = None
        self.root = None

    def insert(self,v):
        #v: insert node
        #z: transition node
        z = self.root
        if z == None:
            # tree: {key:[left child, right chile, parent]}
            self.tree = {v:[None,None,None]}
            self.root = v
        else:
            while z != None:
                #y: parent node
                y = z
                if v < z:
                    flg = 0
                    z = self.tree[z][flg]
                else:
                    flg = 1
                    z = self.tree[z][flg]

            self.tree[y][flg] = v
            # insert node v to the tree
            self.tree.update({v:[None,None,y]})

    def inorder(self,z):
        if z == None:
            return
        self.inorder(self.tree[z][0])
        #print "{}".format(z),
        sys.stdout.write(" "+str(z))
        self.inorder(self.tree[z][1])

    def preorder(self,z):
        if z == None:
            return
        #print "{}".format(z),
        sys.stdout.write(" "+str(z))
        self.preorder(self.tree[z][0])
        self.preorder(self.tree[z][1])

    def find(self,v):
        ans = v in self.tree.keys()
        if ans:
            print "yes"
        else:
            print "no"

    def delete(self,v):
        # v: delete node
        if self.root == None or v not in self.tree.keys():
            print "cannot delete, the node:{} is not included in the tree".format(v)
            return
        # parent list is not "list", but parent node.
        parent_list = self.tree[v][2]

        # when v is the root
        if parent_list == None:
            ## when v has no child
            if self.tree[v][:2] == [None,None]:
                # clear tree
                self.root = None
            ## when v has two children
            elif None not in self.tree[v][:2]:
                # go: transition node
                # tmp: change node
                go = self.tree[v][1]
                while go != None:
                    tmp = go
                    go = self.tree[go][0]
                # set root
                self.root = tmp
                # set tmp to a parent
                self.tree[self.tree[v][0]][2] = tmp
                self.tree[self.tree[v][1]][2] = tmp
                # set the left child of the parent of tmp to None
                self.tree[self.tree[tmp][2]][0] = None
                # set the left child of tmp to the left child of v
                self.tree[tmp][0] = self.tree[v][0]
                # when tmp is not the right child of v                
                if self.tree[v][1] != tmp:
                    right_child = self.tree[tmp][1]
                    self.tree[self.tree[tmp][2]][0] = right_child
                    if right_child != None:
                        self.tree[right_child][2] = self.tree[tmp][2]
                    self.tree[tmp][1] = self.tree[v][1]
                self.tree[tmp][2] = parent_list
            ## when v has a left child, but not has a child child
            elif self.tree[v][1] == None:
                self.root = self.tree[v][0]
                self.tree[self.root][2] = None
            ## when v has a right child, but not has a left child
            else:
                self.root = self.tree[v][1]
                self.tree[self.root][2] = None

        # when v is not the root
        ## when v has no child
        elif self.tree[v][:2] == [None,None]:
            if self.tree[parent_list][0] == v:
                self.tree[parent_list][0] = None
            else:
                self.tree[parent_list][1] = None
        ## when v has two children
        elif None not in self.tree[v][:2]:
            # go: transition node
            # tmp: change node
            go = self.tree[v][1]
            while go != None:
                tmp = go
                go = self.tree[go][0]
            # when v is a left child
            if self.tree[parent_list][0] == v:
                self.tree[parent_list][0] = tmp
                self.tree[self.tree[v][0]][2] = tmp
                self.tree[self.tree[v][1]][2] = tmp

                self.tree[tmp][0] = self.tree[v][0]
                # when tmp is not the right child of v
                if self.tree[v][1] != tmp:
                    right_child = self.tree[tmp][1]
                    self.tree[self.tree[tmp][2]][0] = right_child
                    if right_child != None:
                        self.tree[right_child][2] = self.tree[tmp][2]
                    self.tree[tmp][1] = self.tree[v][1]

                self.tree[tmp][2] = parent_list
            # when v is a right child
            else:
                self.tree[parent_list][1] = tmp
                self.tree[self.tree[v][0]][2] = tmp
                self.tree[self.tree[v][1]][2] = tmp
                # when tmp is not the right child of v
                self.tree[tmp][0] = self.tree[v][0]
                if self.tree[v][1] != tmp:
                    right_child = self.tree[tmp][1]
                    self.tree[self.tree[tmp][2]][0] = right_child
                    if right_child != None:
                        self.tree[right_child][2] = self.tree[tmp][2]
                    self.tree[tmp][1] = self.tree[v][1]

                self.tree[tmp][2] = parent_list
        elif self.tree[v][1] == None:
            if self.tree[parent_list][0] == v:
                self.tree[parent_list][0] = self.tree[v][0]
                self.tree[self.tree[v][0]][2] = parent_list
            else:
                self.tree[parent_list][1] = self.tree[v][0]
                self.tree[self.tree[v][0]][2] = parent_list
        else:
            if self.tree[parent_list][0] == v:
                self.tree[parent_list][0] = self.tree[v][1]
                self.tree[self.tree[v][1]][2] = parent_list
            else:
                self.tree[parent_list][1] = self.tree[v][1]
                self.tree[self.tree[v][1]][2] = parent_list
        del self.tree[v]


if __name__ == '__main__':
    i = 0
    bt = BinaryTree()
    n = int(raw_input())
    while i < n:
        i += 1
        #flg = i < n
        #print flg,i,n
        a = raw_input()
        a = a.split(" ")
        #print a[0],a[1]
        if a[0] == "insert":
            bt.insert(int(a[1]))
        elif a[0] == "print":
            #print "",
            #print ""
            bt.inorder(bt.root)
            print ""
            #print "",
            bt.preorder(bt.root)
            print ""
        elif a[0] == "find":
            bt.find(int(a[1]))
        elif a[0] == "delete":
            bt.delete(int(a[1]))
        else:
            print "error input"
